src/polymer/polymer2-mixin-scanner.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * @license
+ * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as estree from 'estree';
+
+import {Visitor} from '../javascript/estree-visitor';
+import * as esutil from '../javascript/esutil';
+import {JavaScriptDocument} from '../javascript/javascript-document';
+import {JavaScriptScanner} from '../javascript/javascript-scanner';
+import * as jsdoc from '../javascript/jsdoc';
+
+import {ScannedPolymerElementMixin} from './polymer-element-mixin';
+import {getConfig, getProperties} from './polymer2-config';
+
+export class Polymer2MixinScanner implements JavaScriptScanner {
+  async scan(
+      document: JavaScriptDocument, visit: (visitor: Visitor) => Promise<void>):
+      Promise<ScannedPolymerElementMixin[]> {
+    let visitor = new MixinVisitor(document);
+    await visit(visitor);
+    return visitor._mixins;
+  }
+}
+
+class MixinVisitor implements Visitor {
```

Comment:
I think we should be generous with warnings here for mixins that are explicitly declared but that we can't analyze. We could have an additional annotation that indicates that the analyzer won't be able to figure out a more complex mixin, and that it's the user's responsibility at that point to explain it to us.

If we don't, we're in danger of being opaque, mysterious and thus frustrating to use.